Sources/AsyncHTTPClient/ConnectionPool/HTTPConnectionPool+Factory.swift

```diff
@@ -0,0 +1,314 @@
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    private func makePlainBootstrap(eventLoop: EventLoop) -> NIOClientTCPBootstrapProtocol {
+        #if canImport(Network)
+            if #available(OSX 10.14, iOS 12.0, tvOS 12.0, watchOS 6.0, *), let tsBootstrap = NIOTSConnectionBootstrap(validatingGroup: eventLoop) {
+                return tsBootstrap
+                    .addTimeoutIfNeeded(self.clientConfiguration.timeout)
+                    .channelInitializer { channel in
+                        do {
+                            try channel.pipeline.syncOperations.addHandler(HTTPClient.NWErrorHandler())
+                            return channel.eventLoop.makeSucceededFuture(())
+                        } catch {
+                            return channel.eventLoop.makeFailedFuture(error)
+                        }
+                    }
+            }
+        #endif
+
+        if let nioBootstrap = ClientBootstrap(validatingGroup: eventLoop) {
+            return nioBootstrap
+                .addTimeoutIfNeeded(self.clientConfiguration.timeout)
+        }
+
+        preconditionFailure("No matching bootstrap found")
+    }
+
+    private func makeTLSChannel(eventLoop: EventLoop, logger: Logger) -> EventLoopFuture<(Channel, String?)> {
+        let bootstrapFuture = self.makeTLSBootstrap(
+            eventLoop: eventLoop,
+            logger: logger
+        )
+
+        var channelFuture = bootstrapFuture.flatMap { bootstrap -> EventLoopFuture<Channel> in
+            switch self.key.scheme {
+            case .https:
+                return bootstrap.connect(host: self.key.host, port: self.key.port)
+            case .https_unix:
+                return bootstrap.connect(unixDomainSocketPath: self.key.unixPath)
+            case .http, .http_unix, .unix:
+                preconditionFailure("Unexpected schema")
+            }
+        }.flatMap { channel -> EventLoopFuture<(Channel, String?)> in
+            let tlsEventHandler = try! channel.pipeline.syncOperations.handler(type: TLSEventsHandler.self)
+            return tlsEventHandler.tlsEstablishedFuture.flatMap { negotiated in
+                channel.pipeline.removeHandler(tlsEventHandler).map { (channel, negotiated) }
+            }
+        }
+
+        #if canImport(Network)
+            // If NIOTransportSecurity is used, we want to map NWErrors into NWPOsixErrors or NWTLSError.
+            channelFuture = channelFuture.flatMapErrorThrowing { error in
+                throw HTTPClient.NWErrorHandler.translateError(error)
+            }
+        #endif
+
+        return channelFuture
+    }
+
+    private func makeTLSBootstrap(eventLoop: EventLoop, logger: Logger)
+        -> EventLoopFuture<NIOClientTCPBootstrapProtocol> {
+        // since we can support h2, we need to advertise this in alpn
+        var tlsConfig = self.tlsConfiguration
+        tlsConfig.applicationProtocols = ["http/1.1" /* , "h2" */ ]
```

Comment:
Just to confirm, h2 is coming in another PR? (apologies if you've said this somewhere and I missed it)

Reply:
Yes, it is coming in one of the follow up PRs (there will be a lot of them)